subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt

```diff
@@ -0,0 +1,169 @@
+                && !insideImport
+                && !partialDotReferenceExpression
+                && element.text.startsWith(importElement.referencedClass + ".")
+        }
+
+        private val KtImportDirective.referencedClass: String
+            get() = if (aliasName == null) {
+                children.last().children[1].text
+            } else {
+                aliasName!!
+            }
+    }
+
+    private fun findMergedRImport(file: KtFile, app: AndroidModule): KtImportDirective? {
+        val imports: List<KtImportDirective> = file.importList?.imports?.toList() ?: return null
+        return imports.firstOrNull {
+            it.importPath.toString().startsWith(app.packageId)
+        }
+    }
+
+    private fun findUniqueResources(modules: List<AndroidModule>): List<ResourceId> {
+        val uniqueIds = modules
+            .flatMap { it.ids }
+            .unique()
+
+        return uniqueIds.map { id ->
+            val module = requireNotNull(modules.find { module ->
+                module.ids.contains(id)
+            })
+            ResourceId(id, module)
+        }
+    }
+
+    private fun <E> Collection<E>.unique(): Collection<E> {
+        val map = mutableMapOf<E, Int>()
+        this.forEach { element ->
+            val count = map[element]
+            if (count == null) {
+                map[element] = 1
+            } else {
+                map[element] = count + 1
+            }
+        }
+        return map.filter { it.value == 1 }.keys
+    }
+
+    private fun fixImports(file: KtFile, modules: Set<AndroidModule>) {
+        val newImports: List<KtImportDirective> = modules.map {
+            createImportForModule(it)
+        }
+        val importList = requireNotNull(file.children.find { it is KtImportList })
+        newImports.forEach { newImport ->
+            val imports = importList.children.filterIsInstance<KtImportDirective>()
+            if (!imports.contains(newImport)) {
+                importList.add(createNewLineElement())
+                importList.add(newImport)
+            }
+        }
+    }
+
+    private fun createImportForModule(module: AndroidModule): KtImportDirective {
+        val alias = importAliasForResource(module)
+
+        val file = compiler.createKtFile("import ${module.packageId}.R as $alias")
+        return file.children.find { it is KtImportList }!!.children[0] as KtImportDirective
+    }
+
+    private fun createNewLineElement(): PsiWhiteSpace {
+        val file = compiler.createKtFile(
+            """
+
+            import R
+            
+            """.trimIndent()
+        )
+        return file.children.find { it is PsiWhiteSpace }!! as PsiWhiteSpace
+    }
+
+    private fun createElementForResourceRef(resource: ResourceId): PsiElement {
+        val alias = importAliasForResource(resource.module)
+
+        val file = compiler.createKtFile("val id = $alias.id.${resource.id}")
+        return file.children.find { it is KtProperty }!!.children[0]
```

Comment:
Sad that we have to create a whole file.
Do you dig about that?

Reply:
Have tried PsiElementFactory and PsiJavaParserFacade with no luck.
Maybe miss something but don't want to spend more time for this optimization.

Reply:
Here is a starting point: 
https://jetbrains.org/intellij/sdk/docs/basics/architectural_overview/modifying_psi.html?search=psiWHitespace#creating-the-new-psi

Reply:
Fixed here: https://github.com/avito-tech/avito-android/pull/514/files